msgpack-core/src/main/java/org/msgpack/core/buffer/MessageBuffer.java

```diff
@@ -86,20 +88,24 @@
                 directByteBufferConstructor = directByteBufferClass.getDeclaredConstructor(long.class, int.class, Object.class);
             }
             catch(NoSuchMethodException e) {
+                // https://android.googlesource.com/platform/libcore/+/master/luni/src/main/java/java/nio/DirectByteBuffer.java
+                // DirectByteBuffer(long address, int capacity)
                 directByteBufferConstructor = directByteBufferClass.getDeclaredConstructor(long.class, int.class);
-                isAcceptReference = true;
+                isAcceptReference = false;
             }
             byteBufferConstructor = directByteBufferConstructor;
             isByteBufferConstructorTakesBufferReference = isAcceptReference;
             if(byteBufferConstructor == null)
                 throw new RuntimeException("Constructor of DirectByteBuffer is not found");
+
             byteBufferConstructor.setAccessible(true);
 
             // Check the endian of this CPU
             boolean isLittleEndian = true;
             byte[] a = new byte[8];
+            // use Unsafe.putLong to an array since Unsafe.getByte is not available in Android
             unsafe.putLong(a, (long) ARRAY_BYTE_BASE_OFFSET, 0x0102030405060708L);
```

Comment:
Maybe. ByteOrder.nativeOrder is doing almost the same thing:  
{code} 
    static ByteOrder byteOrder() {
	if (byteOrder != null)
	    return byteOrder;
	long a = unsafe.allocateMemory(8);
	try {
	    unsafe.putLong(a, 0x0102030405060708L);
	    byte b = unsafe.getByte(a);
	    switch (b) {
	    case 0x01: byteOrder = ByteOrder.BIG_ENDIAN;     break;
	    case 0x08: byteOrder = ByteOrder.LITTLE_ENDIAN;  break;
	    default:
		throw new Error("Unknown byte order");
	    }
	} finally {
	    unsafe.freeMemory(a);
	}
	return byteOrder;
    }
{code}

But I'm not sure this method are properly implemented for all platforms.